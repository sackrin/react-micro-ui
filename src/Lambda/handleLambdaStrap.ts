import path from 'path';
import { createElement } from 'react';
import embedComponent from '../Helpers/embedComponent';
import { MicroUiConfigProfileEnv } from '../Types/MicroUiConfigProfileEnv';
import { MicroUiConfig } from '../Types/MicroUiConfig';
import { LambdaResponse } from '../Types/LambdaResponse';

type HandleLambdaStrap = (
  name: string,
  component: string,
  logger: any,
  env: MicroUiConfigProfileEnv,
  config: MicroUiConfig,
  method: string,
) => (event: any, context: any) => Promise<LambdaResponse>;

// Direct Import React
// We have to do it this way to permit SSR react + hooks
const ReactDOMServer = require(path.join(process.cwd(), 'node_modules', 'react-dom', 'server'));

const handleLambdaStrap: HandleLambdaStrap = (name, component, logger, env, config, method) => async (
  event,
  context,
) => {
  // Extract and construct the component props from the provided POST and GET variables
  const { queryStringParameters, body } = event;
  const httpQuery = !queryStringParameters ? {} : queryStringParameters;
  const httpBody = !body ? {} : JSON.parse(body);
  const props = { ...httpQuery, ...httpBody };
  // Attempt to construct a static HTML representation of the component
  // Do this using provided POST and GET parameters
  try {
    // Attempt to render the component's HTML using react's ReactDOMServer
    const rendered = ReactDOMServer.renderToString(createElement(component, props));
    const html = embedComponent(name, env, config, props, rendered);
    // Return the static HTML to the client
    return {
      headers: {
        'content-type': 'text/html',
        expires: '-1',
        'cache-control': 'private, no-cache, no-store, must-revalidate',
        pragma: 'no-cache',
      },
      body: html,
      statusCode: 200,
    };
  } catch (e) {
    // Log the error which the renderer encountered
    // @TODO should the error generator be a callback that can be provided to allow better customisation?
    logger.error({ type: 'COMPONENT_RENDER_ERROR', component: name, message: e.message, props });
    // Return a simple error message to the client
    // @TODO should this contain more information?
    // @TODO should this be within the config?
    return {
      headers: {
        'content-type': 'text/html',
        expires: '-1',
        'cache-control': 'private, no-cache, no-store, must-revalidate',
        pragma: 'no-cache',
      },
      body: 'Component Error',
      statusCode: 500,
    };
  }
};

export default handleLambdaStrap;
