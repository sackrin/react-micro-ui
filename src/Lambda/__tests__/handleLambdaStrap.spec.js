import React from 'react';
import { expect } from 'chai';
import handleLambdaStrap from '../handleLambdaStrap';
import config from '../../microui.default.config';

describe('Lambda/handleLambdaStrap', () => {
  const mockContext = {};
  const ExampleComponent = props => <div>{props.name}</div>;

  it('can correctly out a server side rendered component with GET variables', async () => {
    const mockEvent = {
      queryStringParameters: { name: 'John Cater' },
      body: JSON.stringify({})
    };
    const payload = await handleLambdaStrap('ExampleComponent', ExampleComponent, console, {}, config, 'GET')(mockEvent, mockContext);
    jestExpect(payload).toMatchSnapshot();
  });

  it('can correctly out a server side rendered component with POST variables', async () => {
    const mockEvent = {
      queryStringParameters: {},
      body: JSON.stringify({ name: 'John Cater' })
    };
    const payload = await handleLambdaStrap('ExampleComponent', ExampleComponent, console, {}, config, 'GET')(mockEvent, mockContext);
    jestExpect(payload).toMatchSnapshot();
  });

});
