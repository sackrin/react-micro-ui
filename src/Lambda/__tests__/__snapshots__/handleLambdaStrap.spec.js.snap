// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lambda/handleLambdaStrap can correctly out a server side rendered component with GET variables 1`] = `
Object {
  "body": "<div data-microui-library=\\"exampleMicroUI\\" data-microui-component=\\"ExampleComponent\\"><div data-reactroot=\\"\\">John Cater</div></div><script type=\\"application/javascript\\"> (function(n, c, p, t) {  var w = window, m = function(e){   if (e.detail.name === n && w[n] && w[n].Hydrate) {    w[n].Hydrate(document.querySelector(\`div[data-microui-component=\\"ExampleComponent\\"] div\`),c,p,t);   };  };  if (w[\`exampleMicroUI\`]) { m(); } else { w.addEventListener('microUILoaded', m); }  var d = document.getElementById(\`exampleMicroUILibrary\`);  if (d === null || d.length === 0) {   var tag = document.createElement('script');   tag.id = \`exampleMicroUILibrary\`;   tag.src = \`http://localhost:9000/bootstrap.js\`;   document.body.appendChild(tag);  } })('exampleMicroUI', 'ExampleComponent', {\\"name\\":\\"John Cater\\",\\"apiUrl\\":\\"http://localhost:9000\\",\\"apiPath\\":\\"/api/v1\\",\\"assetUrl\\":\\"http://localhost:9000\\",\\"assetTarget\\":\\"umd\\"}, {});</script>",
  "headers": Object {
    "cache-control": "private, no-cache, no-store, must-revalidate",
    "content-type": "text/html",
    "expires": "-1",
    "pragma": "no-cache",
  },
  "statusCode": 200,
}
`;

exports[`Lambda/handleLambdaStrap can correctly out a server side rendered component with POST variables 1`] = `
Object {
  "body": "<div data-microui-library=\\"exampleMicroUI\\" data-microui-component=\\"ExampleComponent\\"><div data-reactroot=\\"\\">John Cater</div></div><script type=\\"application/javascript\\"> (function(n, c, p, t) {  var w = window, m = function(e){   if (e.detail.name === n && w[n] && w[n].Hydrate) {    w[n].Hydrate(document.querySelector(\`div[data-microui-component=\\"ExampleComponent\\"] div\`),c,p,t);   };  };  if (w[\`exampleMicroUI\`]) { m(); } else { w.addEventListener('microUILoaded', m); }  var d = document.getElementById(\`exampleMicroUILibrary\`);  if (d === null || d.length === 0) {   var tag = document.createElement('script');   tag.id = \`exampleMicroUILibrary\`;   tag.src = \`http://localhost:9000/bootstrap.js\`;   document.body.appendChild(tag);  } })('exampleMicroUI', 'ExampleComponent', {\\"name\\":\\"John Cater\\",\\"apiUrl\\":\\"http://localhost:9000\\",\\"apiPath\\":\\"/api/v1\\",\\"assetUrl\\":\\"http://localhost:9000\\",\\"assetTarget\\":\\"umd\\"}, {});</script>",
  "headers": Object {
    "cache-control": "private, no-cache, no-store, must-revalidate",
    "content-type": "text/html",
    "expires": "-1",
    "pragma": "no-cache",
  },
  "statusCode": 200,
}
`;
